src/ZkFold/Base/Protocol/Protostar/Fold.hs

```diff
@@ -0,0 +1,179 @@
+{-# LANGUAGE AllowAmbiguousTypes  #-}
+{-# LANGUAGE DeriveAnyClass       #-}
+{-# LANGUAGE TypeOperators        #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+module ZkFold.Base.Protocol.Protostar.Fold where
+
+
+import           Control.DeepSeq                                     (NFData)
+import           Control.Lens                                        ((^.))
+import           Data.Map.Strict                                     (Map)
+import qualified Data.Map.Strict                                     as M
+import           GHC.Generics                                        (Generic)
+import           Prelude                                             (and, otherwise, type (~), ($), (<$>), (<*>), (==))
+import qualified Prelude                                             as P
+
+import           ZkFold.Base.Algebra.Basic.Class
+import           ZkFold.Base.Algebra.Basic.Number
+import qualified ZkFold.Base.Algebra.Polynomials.Univariate          as PU
+import           ZkFold.Base.Data.Vector                             (Vector)
+import           ZkFold.Base.Protocol.Protostar.Accumulator
+import qualified ZkFold.Base.Protocol.Protostar.AccumulatorScheme    as Acc
+import           ZkFold.Base.Protocol.Protostar.ArithmeticCircuit
+import           ZkFold.Base.Protocol.Protostar.Commit
+import           ZkFold.Base.Protocol.Protostar.CommitOpen
+import           ZkFold.Base.Protocol.Protostar.FiatShamir
+import           ZkFold.Base.Protocol.Protostar.Oracle
+import qualified ZkFold.Base.Protocol.Protostar.SpecialSound         as SPS
+import           ZkFold.Prelude                                      (replicate)
+import           ZkFold.Symbolic.Compiler
+import           ZkFold.Symbolic.Compiler.ArithmeticCircuit.Internal
+import           ZkFold.Symbolic.Data.FieldElement                   (FieldElement)
+
+data FoldResult n c a
+    = FoldResult
+    { output          :: Vector n a
+    , lastAccumulator :: Accumulator (Vector n a) a c (Map Natural a)
+    , verifierOutput  :: P.Bool
+    , deciderOutput   :: P.Bool
+    } deriving (P.Show, Generic, NFData)
+
+type FS_CM n c a = FiatShamir a (CommitOpen a c (RecursiveCircuit n a))
+
+transform
+    :: forall n c a
+    .  HomomorphicCommit a [Map Natural a] c
```

Comment:
These look like references to concrete instances, maybe it is better to inline them then?